src/Init/System/IO.lean

```diff
@@ -210,14 +213,25 @@ def Handle.putStr (h : Handle) (s : String) : m Unit :=
 def Handle.putStrLn (h : Handle) (s : String) : m Unit :=
   h.putStr (s.push '\n')
 
--- TODO: support for binary files
-partial def Handle.readToEnd (h : Handle) : m String :=
+partial def Handle.readBinToEnd (h : Handle) : m ByteArray := do
+  let rec loop (left : Nat) : m ByteArray := do
+    if left == 0 then ByteArray.empty
+    else
+      let buf ← h.read left
+      let left := left - buf.size
+      pure $ buf ++ (←loop left)
+  loop (←h.size)
```

Comment:
I don't think we should rely on `h.size` here; it could get out of date while reading, or not even exist for some kinds of files. We could potentially use it as an initial size for the buffer, but that's a minor optimization.

Reply:
Yes, this implementation is quite bad and I only added it to make some tests work. We could have a `bytesTillEnd` method but that still won't help with e.g. pipes. To write it properly, I'd port over `read_to_end` from Rust. For now though, I guess we can just try to read `256` bytes at a time or something of that order.